File: snippets/0743_network_delay_time.py
from collections import defaultdict
from heapq import heappop, heappush
from typing import List


# 1. Dijkstra - Set
def networkDelayTime1(times: List[List[int]], n: int, k: int) -> int:
    graph = defaultdict(list)
    for u, v, w in times:
        graph[u].append((v, w))

    minHeap = [(0, k)]  # (weight, node)
    visited = set()
    delay = 0

    while minHeap:
        cur, u = heappop(minHeap)

        if u in visited:
            continue
        visited.add(u)

        delay = max(delay, cur)

        for v, w in graph[u]:
            if v not in visited:
                heappush(minHeap, (cur + w, v))

    return delay if len(visited) == n else -1


# 2. Dijkstra - Dict
def networkDelayTime2(times: List[List[int]], n: int, k: int) -> int:
    graph = defaultdict(list)
    for u, v, w in times:
        graph[u].append((v, w))

    minHeap = [(0, k)]  # (weight, node)
    dist = defaultdict(int)
    dist[k] = 0

    while minHeap:
        cur, u = heappop(minHeap)

        if cur > dist[u]:
            continue

        for v, w in graph[u]:
            path = cur + w
            if v not in dist or path < dist[v]:
                dist[v] = path
                heappush(minHeap, (path, v))

    return max(dist.values()) if len(dist) == n else -1


# 3. Bellman-Ford
def networkDelayTimeBF(times: List[List[int]], n: int, k: int) -> int:
    # TC: O(N * E)
    # SC: O(N)

    dist = [float("inf")] * n
    dist[k - 1] = 0

    for _ in range(n - 1):
        for u, v, w in times:
            if dist[u - 1] + w < dist[v - 1]:
                dist[v - 1] = dist[u - 1] + w

    if any(d == float("inf") for d in dist):
        return -1

    return max(dist)


times = [[2, 1, 1], [2, 3, 1], [3, 4, 1]]
n = 4
k = 2
print(networkDelayTime1(times, n, k))  # 2
print(networkDelayTime2(times, n, k))  # 2
print(networkDelayTimeBF(times, n, k))  # 2

File: leetpattern/cpp/0601_0900/0743_network_delay_time.cc
#include <algorithm>
#include <cassert>
#include <climits>
#include <queue>
#include <vector>
using namespace std;

class NetworkDelayTime {
   public:
    // single-source, non-negative weight -> Dijkstra's
    // https://leetcode.cn/problems/network-delay-time/solutions/2668220/liang-chong-dijkstra-xie-fa-fu-ti-dan-py-ooe8/
    int dijkstra(vector<vector<int>>& times, int n, int k) {
        vector<vector<pair<int, int>>> graph(n + 1);
        for (auto& t : times) {
            graph[t[0]].push_back({t[1], t[2]});
        }

        vector<int> dist(n + 1, INT_MAX);
        // min heap
        priority_queue<pair<int, int>, vector<pair<int, int>>,
                       greater<pair<int, int>>>
            min_heap;

        dist[k] = 0;
        min_heap.push({0, k});  // [dist, node]

        while (!min_heap.empty()) {
            auto [d, u] = min_heap.top();
            min_heap.pop();

            if (d > dist[u]) continue;  // found the shortest

            for (auto& [v, w] : graph[u]) {
                if (dist[u] + w < dist[v]) {
                    dist[v] = dist[u] + w;
                    min_heap.push({dist[v], v});
                }
            }
        }

        int res = 0;
        for (int i = 1; i <= n; ++i) {
            if (dist[i] == INT_MAX) return -1;
            res = max(res, dist[i]);
        }
        return res;
    }

    // single source -> Bellman-Ford
    int bellman_ford(vector<vector<int>>& times, int n, int k) {
        vector<int> dist(n + 1, INT_MAX);
        dist[k] = 0;

        for (int i = 1; i <= n - 1; i++) {
            for (auto& edge : times) {
                int u = edge[0], v = edge[1], t = edge[2];
                if (dist[u] != INT_MAX && dist[v] > dist[u] + t) {
                    dist[v] = dist[u] + t;
                }
            }
        }

        int res = 0;
        for (int i = 1; i <= n; i++) {
            if (dist[i] == INT_MAX) return -1;
            res = max(res, dist[i]);
        }
        return res;
    }
};

int main() {
    NetworkDelayTime solution;
    vector<vector<int>> times = {{2, 1, 1}, {2, 3, 1}, {3, 4, 1}};
    assert(solution.dijkstra(times, 4, 2) == 2);
    assert(solution.bellman_ford(times, 4, 2) == 2);

    return 0;
}
